### PR TITLE
file

### LINKED ISSUE #132: File input
For cases where my dataset is not available online and I don’t want to upload it, it would be awesome to have https://observablehq.com/@mbostock/file-input, ideally with at least text, JSON, and CSV variants. It would also be great to have an option to enable or disable the URL input.

### PR DESCRIPTION
**Pull request by mbostock**
Fixes #132. Based on https://observablehq.com/@mbostock/localfile. It looks like this:

Screen Shot 2021-09-30 at 10 25 22 AM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -349,6 +350,40 @@ The available *options* are:
...
+
+#### Inputs.textarea(*options*)
+
+[<img src="./img/textarea.png" alt="A Textarea asking for your biography" width="660">](https://observablehq.com/@observablehq/input-textarea)
+
+```js
+viewof bio = Inputs.textarea({label: "Biography", placeholder: "Tell us a little about yourself…"})
+```
+
+[Source](./src/textarea.js) · [Examples](https://observablehq.com/@observablehq/input-textarea) · A Textarea allows multi-line freeform text input. By default, a Textarea will report its value immediately on input. If more deliberate behavior is desired, the *submit* option can be set to true to wait until a button is clicked or Command-Enter is pressed.
```
Comment (visnup) [inline README.md:363]: Command/Ctrl-Enter? I'm assuming ctrl-enter works on non-Mac?

File: README.md [hunk 2]
```diff
@@ -374,39 +409,33 @@ The available *options* are:
...
-* *spellcheck* - whether to activate the browser’s spell-checker.
-* *autocomplete* - the [autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) attribute, as text or boolean (true for on, false for off).
-* *autocapitalize* - the [autocapitalize](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autocapitalize) attribute, as text or boolean (true for on, false for off).
-* *minlength* - [minimum length](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/minlength) attribute.
-* *maxlength* - [maximum length](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/maxlength) attribute.
-* *required* - if true, the input must be non-empty; defaults to *minlength* > 0.
+* *required* - if true, the input must be a valid date.
 * *validate* - a function to check whether the text input is valid.
+* *accept* - the [acceptable file types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#accept).
+* *capture* - for [capturing image or video data](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#capture).
```
Comment (visnup) [inline README.md:431]: TIL

File: src/file.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+export function fileOf(AbstractFile) {
+
+  class LocalFile extends AbstractFile {
+    constructor(file) {
+      super(file.name);
+      Object.defineProperty(this, "_", {value: file});
+      Object.defineProperty(this, "_url", {writable: true});
+    }
+    async url() {
+      return this._url || (this._url = URL.createObjectURL(this._));
```
Comment (visnup) [inline src/file.js:15]: Nice.

**Review by visnup — APPROVED** [review]
This input makes me somewhat wonder if there's a UX workflow that "promotes" the locally uploaded file into a file attachment. Though I guess this also is another form of a more generic "snapshot this cell into a file attachment" or simply "snapshot this cell"…

File: README.md (hunk 1 above)
Comment (mbostock) [reply README.md:363]: I didn’t change this documentation for this PR (I just moved it around), but yeah, it is assuming macOS here.

File: src/file.js (hunk 3 above)
Comment (mbostock) [reply src/file.js:15]: Nice.

**Pull request merged**